### ISSUE TITLE
TLS 1.3 support 

### ISSUE THREAD

**Issue by x0r2d2** [angristan/openvpn-install#637]
Hello to dev team!

Just noticed that Openvpn Connect for Android now supports tls 1.3 ciphers (still beta).

`23:53:28.237 -- SSL Handshake: CN=server_icij7YmhfXXXXXXX, TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384`

Openvpn for Android also supports tls 1.3. I think all other apps also support tls 1.3 with openssl 1.1.1c.

![IMG_20200426_000821](https://user-images.githubusercontent.com/12900872/80288550-16ee3780-8752-11ea-8f50-82d7b93814d6.jpg)

Maybe we can update config to use tls 1.3 only?

**Comment by randshell** [comment]
That would be great but it requires some rewrite:
> - There are new ciphersuites that only work in TLSv1.3. The old ciphersuites cannot be used for TLSv1.3 connections and the new ones cannot be used in TLSv1.2 and below.
> - The new ciphersuites are defined differently and do not specify the certificate type (e.g. RSA, DSA, ECDSA) or the key exchange mechanism (e.g. DHE or ECHDE). This has implications for ciphersuite configuration.
> - DSA certificates are no longer allowed in TLSv1.3 connections

Source: [OpenSSL wiki](https://wiki.openssl.org/index.php/TLS1.3)

> use tls 1.3 only

@hybtoy I don't think we can yet. Changing the config to TLS 1.3 breaks compatibility with Ubuntu 16.04, Centos 7 and Debian 9 because they don't have openssl 1.1.1 in the repositories.

Adding support for 1.3 and not dropping 1.2 in the same script adds some code and test cases... @angristan if you're ok with this or maybe we can use another branch for 1.3 only like we did for openvpn 2.4?

**Comment by TinCanTech** [comment]
Openvpn will provide TLS 1.3 as it becomes available like so:
`Sat Apr 25 23:01:18 2020 us=464971 tct.66.c.ub18/ip.add.re.ss:3455 Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, 384 bit EC, curve: secp384r1`

**Comment by randshell** [comment]
@TinCanTech Github saves previous comments where you pasted your IP. You may want to delete them https://help.github.com/en/enterprise/2.19/user/github/building-a-strong-community/tracking-changes-in-a-comment

**Comment by TinCanTech** [comment]
It is a mutable client IP .. of no interest to anybody.  I simply cleaned up the comment. But thanks for the pointer !

**Comment by randshell** [comment]
> Openvpn will provide TLS 1.3 as it becomes available like so:
`Sat Apr 25 23:01:18 2020 us=464971 tct.66.c.ub18/ip.add.re.ss:3455 Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, 384 bit EC, curve: secp384r1`


Thanks @TinCanTech 

So there isn't a need to explicit this configuration in the script now. The renegotiation handles it and uses a secure default already.

**Comment by TinCanTech** [comment]
Just to add a little clarity.  OpenVPN will select the best/highest (what ever term you use to describe the security level) between peers.

For example: 

- Server running Linux may have `openvpn` built with `openssl 1.1.1f`
- Client running Windows may have `openvpn` built with `openssl 1.1.0j`

`openvpn` will select the best **matching** TLS settings.

FTR: The above scenario results in:

`Control Channel: TLSv1.2, cipher TLSv1.2 ECDHE-ECDSA-AES256-GCM-SHA384, 384 bit EC, curve: secp384r1`

Had the server been configured to use TLS 1.3 then this client would not be able to connect.

**Comment by randshell** [comment]
> openvpn will select the best matching TLS settings.

Maybe we can default to `TLS_CHACHA20_POLY1305_SHA256` in the script instead of `TLS_AES_256_GCM_SHA384` for performance? It benefits on mobile devices, routers and raspberrys and it's still the second most secure option.

**Comment by TinCanTech** [comment]
Personally, if you decide to change the script, I would make it a user choice .. with a FAQ.

**Comment by randshell** [comment]
Yes, of course. I meant the default of the auto completion :smile: 

https://github.com/angristan/openvpn-install/blob/5e2e67f78df9f92a8d7ce92338a8f6e653c0c149/openvpn-install.sh#L394
https://github.com/angristan/openvpn-install/blob/5e2e67f78df9f92a8d7ce92338a8f6e653c0c149/openvpn-install.sh#L401

**Comment by TinCanTech** [comment]
Your two links above relate directly to the data channel not the control channel.

> Maybe we can default to TLS_CHACHA20_POLY1305_SHA256 in the script instead of TLS_AES_256_GCM_SHA384 for performance?

This would have the effect of disabling any user AES-NI .. so I would not make this change.  Hence: a FAQ.

**Comment by perfectpeach** [comment]
Hiya, 



When the OP/Developer will update this script to enable us to select 'TLS 1.3' on installation?

Tnx and best of luck

**Comment by TinCanTech** [comment]
> When the OP/Developer will update this script to enable us to select 'TLS 1.3' on installation?

That will be when:
 * OpenSSL makes this a suitable option
 * OpenVPN implements the changes you are looking for

**Harassing** this project will not produce results.

Also, your avatar is completely **unacceptable**.
